- Add rules for using , naming and casing properties (#8)
zbra-solutions/code-conventions: ZBRA General Code Conventions Guide.adoc

[[never-use-magic-numbers]]
=== Never use 'magic' Numbers

> Never use magic numbers. Never place literal numerical values directly into the source code. Replacing these later on in case of changes is

Author: Realmente a regra quer dizer nunca em lugar algum. portanto source code é adquado. o q falta é alguma coisa tipo "se vc precisar mesmo entao faça assim"

> Never use magic numbers. Never place literal numerical values directly into the source code. Replacing these later on in case of changes is

Reviewer: Poxa, estou perdido aqui, gostaria de entender.. Vai colocar onde? No banco de dados? Onde vou colocar que a "o volume do canister deve ser o dobro (2x) do volume do líquido", por exemplo?

No Banco de dados? Se sim, nós não fazemos isso hoje e me parece que esse tipo de decisão é algo mais de arquitetura de software, não de code guide lines, não?

> Never use magic numbers. Never place literal numerical values directly into the source code. Replacing these later on in case of changes is

Author: O que quis dizer é que é a segunda parte é que está errada, não a primeira.
A palavra operante é "directly", não "source code".  usar um static field é uma forma indireta de usar o numero.

Expressões semelhantes 

"A magic number is a bare naked number used in source code." - [1](https://www.jarchitect.com/QACenter/index.php?qa=155&qa_1=no-magic-numbers)

"In near future we might be enforced by a rule by which we can not have any hard coded numbers in our java source code" -[https://stackoverflow.com/questions/6346421/how-far-should-best-practices-like-avoiding-magic-number-should-go](url)

"source code" é usado no sentido geral. o que é operante é o que vem antes "naked number"  ou "hard coded". No caso eu usei "place directly"

Enumerar todos os usos não ajuda. A ideia é abstrair o que esses usos têm em comum. Mas não ha nenhuma expressão obvia disso. Hard coded por exemplo também é ambiguo. O numero continua hardcoded , apenas em outro lugar. 

A ideia é estabelecer o que é um magic number. Um magic number é literal numérico um numero não identificado, que não tem um nome associado. É difícil definir "identificado" . Aqui vai uma nova tentativa:

_Never use magic numbers. Never place non identified literal numerical values into the source code. Replacing these later on in case of changes is error-prone and unproductive. Instead, always declare them indirectly as values of an immutable static field. Prefer it to be private. 
Choose the name of the field to properly identify the number._

> Never use magic numbers. Never place literal numerical values directly into the source code. Replacing these later on in case of changes is

Reviewer: Essa nova tentativa ficou ótima IMHO.

> error-prone and unproductive. Instead, declare an immutable static field which contains the number. Prefer it to be private. 

Author: field e method sao conceitos basico em OO. algum exemplo da liguagem OO que não tenha field? lembrando que so usamos linguagens OO e se nao for OO estas convencoes nao se aplicam pq nao vai haver metodos nem classes nem um bando de outras coisas também...

> error-prone and unproductive. Instead, declare an immutable static field which contains the number. Prefer it to be private. 

Author: a questao de ser static é simplesmente poupar memoria. um int na classe sao 32 bits. um int em cada instanvia sao 32 bis x numero de instancias. como o valor é sempre o mesmo para todas ser static é mais economico

> error-prone and unproductive. Instead, declare an immutable static field which contains the number. Prefer it to be private. 

Reviewer: É.. eu estava pensando em C ou python... ok.

> Always select the name of the field to inform the role of the number in the code.

Reviewer 2: Concordo com o @ltlombardi.

> Always select the name of the field to inform the role of the number in the code.

Reviewer 3: a ultima linha está kinda duplicada, não?

> Always select the name of the field to inform the role of the number in the code.

Author: Fixed

[source, csharp]
----
// C#

> private static readonly MaximumNumberOfTries = 3;

Author: nao me parece que mudou. a regra de fields com pascal case é para quando sao publicos. o que os nossos nunca são pela nossa convenção de q sejam private. apenas os static public seria dessa forma. mas aqui eu coloquei pascal por causa de ser uma constante e nao pq é public. ai realmente posso ter inventado . gostaria de puvir mais opinioes em relacao ao casing

> private static readonly MaximumNumberOfTries = 3;

Reviewer: É pode ser.. 
Eu acho útil essa diferenciação, de private field ser em camelCase, mas é algo pequeno. Poderíamos padronizar em PascalCase, simplificando.

> private static readonly MaximumNumberOfTries = 3;

Author: Realmente parece que a convenção é usar Pascal Case mesmo quando a constante é privada https://stackoverflow.com/questions/242534/c-sharp-naming-convention-for-constants

In this example `parseInt` is a javaScript global function. The language accepts both syntaxes, but the second syntax does not provide the expected result. It really returns `[1, NaN , 3]` <<ForceLambda>>. In other languages the problem may not be so drastic but forgetting to handle nulls is often a concern easy to overlook when using function references.

[[properties]]
=== Only use properties when operation is not expensive and has no side effects.

> When deciding between using a method with no parameters or a property acessor to obtain some information from an object take attention to the operation expensiveness.

Reviewer 3: ```suggestion
When deciding between using a method with no parameters or a property accessor to obtain some information from an object take attention to the operation expensiveness.
```

Author: Fixed

* When the operation returns in constant time but takes too long so that you want to communicate to the programmer that they should consider caching the result.
* When calling the member twice in succession produces different results (like `Random.next()`)
* When the calling order is important. Note that a property should be able to be set and retrieved in any order.
* When the operation is a conversion, such as `Object.toString`
* When obtaining the property value would have an observable side effect
* When the member is static but returns a value that can be changed (like in a Static Factory Method)
* When you mean to return an array or list. Properties that return arrays and lists can be very misleading. Usually it is necessary to return a copy of the internal array or list so that the user cannot change internal state. This, coupled with the fact that a user can easily assume it is an indexed property, leads to inefficient code. You may use a property when the list is garanteed to be read only or when changing the internal state is not a problem (like when using an object written in the Property Bag pattern) 

[[properties-in-interfaces]]
==== Properties and Interfaces

> Given the rules above, follows that acessors may never be defined in interfaces since the implementation is unkown making it impossible to garantee the rules. For interfaces, methods must always be use instead of properties. On the other hand, property bags should not share interfaces, so this is not a common problem.

Reviewer 3: ```suggestion
Given the rules above, follows that accessors may never be defined in interfaces, as the implementation is unknown and makes it impossible to guarantee the rules. For interfaces, methods must always be used instead of properties. On the other hand, property bags should not share interfaces, so this is not a common problem.
```

Author: Fixed

> Given the rules above, follows that acessors may never be defined in interfaces since the implementation is unkown making it impossible to garantee the rules. For interfaces, methods must always be use instead of properties. On the other hand, property bags should not share interfaces, so this is not a common problem.

Reviewer 3: If the rules itself does not guarantees that, how can the interface do that?
